### ISSUE TITLE
Proposed use of machine options with machine_execute resource

### ISSUE THREAD

**Issue by mwrock** [chef-boneyard/chef-provisioning#340]
Currently the `machine_execute` resource uses the chef server config as the `machine_options` when calling the driver's `connect_to_machine`. I'm guessing the use case here is a user wants to run commands using ssh settings stored in their knife.rb.

I'd like to use this resource but be able to use the same `with_machine_options` semantics that are available with the `machine` resource.

I'd like to propose that we add a `machine_options` attribute to machine_execute that, if provided, would be merged with the chef_server config and the machine_options would override any of the same options in the server config.

I could submit this if this sounds reasonable.

**Comment by jkeiser** [comment]
The main question (and the main reason it's not in there) is--when the node was created, we stuffed information inside about how to connect to it.  Does that information not apply anymore--do we need different information, or need to be able to update it?  What happens when you try to connect?

Part of the reason for the current design is to make friction-free connection possible, so that managing the machines is less of a "try to remember all the options you need" chore.  I realize we're not fully at that goal, so I'm not 100% opposed to workarounds like this in principle, but I'd like to at least record the issues that led us to it so we can get the right information into the node instead of the machine options.

**Comment by mwrock** [comment]
Yeah looking further into this, I see now that is the pattern. I've been using the provisioning for a while now but really just for bootstrapping. I'm now setting out to write "orchestration" recipes where I am moving chef managed infrastructure into legacy datacenters and rolling nonmanaged infrastructure out with 0 downtime. Different pieces of infrastructures require different orchestration. Where I may want to turn off auto convergence, change runlists, invoke a converge, and turn auto convergence back on when I am done and do this all pointing to different nodes throughout the process.

We have not been storing any keys or auth info with the nodes. We also add a recipe to every node's runlist after provisioning that changes its password using a key in the chef vault. I had thought I would populate the machine options with that key and then use these resources but I'll have a look at fiddling with the node data to accomplish the same thing.

I do see a use case for accessing these with machine options. Especially with vsphere which I am working with that does not have a native keystore concept that I know of. I'm trying not to add any code to my driver that is specific to my org. These particular recipes will likely be run on a local workstation and the user would have access to the vault and therefore the auth info needed to get on the box which we may not want to keep on the server in our case.

**Comment by mwrock** [comment]
Just discovered this dirty hack for my recipe:
```
config = node.run_state[:chef_provisioning].config
config[:machine_options] = machine_options
```

**Comment by jkeiser** [comment]
Gotcha!  This requires a better auth handling :/  I've been wanting something like this for a while:

```ruby
Net::SSH.new('55.23.42.123', Chef::Credentials.get('ssh:55.23.42.123'))
```

Or along those lines, anyway.  The SSH transport uses it, and then you just hook into that, implementing with your own handler that grabs from your credentials store.

**Comment by Saburesh07** [comment]
This issue or pull request is being automatically closed because the repository is being archived as part of the [Repository Standardization Initiative](https://github.com/chef-boneyard/oss-repo-standardization-2025).
